examples/misc/seam_carving.py

```diff
@@ -0,0 +1,190 @@
+# /// script
+# requires-python = ">=3.10"
+# dependencies = [
+#     "marimo",
+#     "numba==0.60.0",
+#     "numpy==2.0.2",
+#     "requests==2.32.3",
+#     "scikit-image==0.24.0",
+# ]
+# ///
+
+import marimo
+
+__generated_with = "0.9.6"
+app = marimo.App(width="medium")
+
+
+@app.cell(hide_code=True)
+def __(mo):
+    mo.md(
+        """
+        # Seam Carving 
+
+        _Example adapted from work by [Vincent Warmerdam](https://x.com/fishnets88)_.
+
+        ## The seam carving algorithm
+        This marimo demonstration is partially an homage to [a great video by Grant
+        Sanderson](https://www.youtube.com/watch?v=rpB6zQNsbQU) of 3Blue1Brown, which demonstrates
+        the seam carving algorithm in [Pluto.jl](https://plutojl.org/):
+
+        <iframe width="560" height="315" src="https://www.youtube.com/embed/rpB6zQNsbQU?si=oiZclGIj2atJR47m" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" referrerpolicy="strict-origin-when-cross-origin" allowfullscreen></iframe>
+
+        As Grant explains, the seam carving algorithm preserves the shapes of the main content in the image, while killing the "dead space": the image is resized, but the clocks and other content are not resized or deformed.
+
+        This notebook is a Python version of the seam carving algorithm, but it is also a
+        demonstration of marimo's [persistent caching
+        feature](https://docs.marimo.io/recipes.html#persistent-caching-for-very-expensive-computations),
+        which is helpful because the algorithm is compute intensive even when you
+        use [Numba](https://numba.pydata.org/).
+
+        Try it out by playing with the slider!
+        """
+    )
+    return
+
+
+@app.cell(hide_code=True)
+def __():
+    import requests
+
+    input_image = "The_Persistence_of_Memory.jpg"
+    img_data = requests.get(
+        "https://upload.wikimedia.org/wikipedia/en/d/dd/The_Persistence_of_Memory.jpg"
+    ).content
+
+    with open(input_image, "wb") as handler:
+        handler.write(img_data)
+    return handler, img_data, input_image, requests
+
+
+@app.cell(hide_code=True)
+def __(mo):
+    mo.md("""## Try it!""")
+    return
+
+
+@app.cell
+def __():
+    import marimo as mo
+
+    slider = mo.ui.slider(
+        0.7,
+        1.0,
+        step=0.05,
+        value=1.0,
+        label="Amount of resizing to perform:",
+        show_value=True,
+    )
+    slider
+    return mo, slider
+
+
+@app.cell
+def __(efficient_seam_carve, input_image, mo, slider):
+    with mo.persistent_cache("seam_carves"):
```

Comment:
Doesn't make much difference right now, but with remote caching this would be better:

```python
steps = np.linspace(0.7, 1, 60)
slider = mo.ui.slider(
        steps=steps,
        value=1.0,
        label="Amount of resizing to perform:",
        show_value=True,
    )
    
# Compute them all, so cache hits never happen
with mo.persistent_cache("seam_carves"):
     # Or even do in parallel
     results = list(map(lambda x: efficient_seam_carve(input_image, x), steps))

result = results[np.argmin((steps - slider.value)**2)]
```

Which would allow embedding into a blog post with 0 latency.
I think this is still ideal if you are still developing / messing around though

Reply:
oh my, i really like that idea ✨